### PR TITLE
[app] Add binding support and switch example

### PR DESCRIPTION
**Pull request by gjc13**
#### Problem

Binding cluster support is missing in the SDK.

#### Change overview

*   The change replaces the Ember binding table with one containing both
    bound clusters and peer devices. The new binding table will manage the
    CASE connections to the bound devices.
*   Add switch app  as an example of binding cluster. After binding,
    all the changes on the On/Off cluster will be reflected on the bound
    device.

#### Testing

The PR requires https://github.com/project-chip/connectedhomeip/pull/11963 to work.
https://github.com/project-chip/connectedhomeip/pull/11972 is created to reduce static data usage.

Build & run all-cluster-app and switch-app.

1. Pair all-cluster-app with `chip-tool pairing ble-wifi 1234 $SSID $PSK 0 $PAIRING_CODE $DISCRIMINATOR`
2. Pair switch-app with `chip-tool pairing ethernet 1235 20202021 3840 127.0.0.1 5540`
3. Setup binding with `chip-tool binding bind 1234 0 1 6 1235 1`
4. Toggle the On/Off cluster on the switch app with `chip-tool onoff toggle 1235 1`. The toggle will be also reflected on the all-cluster-app.

### CODE REVIEW

File: examples/switch-app/linux/.gn [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2020 Project CHIP Authors
```
Comment (dhrishi) [inline examples/switch-app/linux/.gn:1]: Nit: 2021 at all places?

File: examples/switch-app/linux/README.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
+# CHIP Linux Lighting Example
```
Comment (dhrishi) [inline examples/switch-app/linux/README.md]: Switch Example

File: examples/switch-app/linux/README.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+An example showing the use of CHIP on the Linux. The document will describe how
+to build and run CHIP Linux Lighting Example on Raspberry Pi. This doc is tested
+on **Ubuntu for Raspberry Pi Server 20.04 LTS (aarch64)** and **Ubuntu for
+Raspberry Pi Desktop 20.10 (aarch64)**
```
Comment (dhrishi) [inline examples/switch-app/linux/README.md]: Please update the documentation as per the switch app

File: examples/switch-app/linux/include/CHIPProjectAppConfig.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+#define CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY 0
+
+// Bulbs do not typically use this - enabled so we can use shell to discover commissioners
+#define CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY_CLIENT 1
+
+#define CHIP_DEVICE_CONFIG_ENABLE_EXTENDED_DISCOVERY 1
+
+#define CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONABLE_DEVICE_TYPE 1
+
+#define CHIP_DEVICE_CONFIG_DEVICE_TYPE 257 // 0x0101 = 257 = Dimmable Bulb
```
Comment (dhrishi) [inline examples/switch-app/linux/include/CHIPProjectAppConfig.h:42]: Should this be 0x0103 -- On/Off Light Switch?

File: examples/switch-app/linux/include/CHIPProjectAppConfig.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+#define CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONABLE_DEVICE_NAME 1
+
+#define CHIP_DEVICE_CONFIG_DEVICE_NAME "Test Bulb"
```
Comment (dhrishi) [inline examples/switch-app/linux/include/CHIPProjectAppConfig.h:46]: Test Switch?

File: examples/switch-app/switch-common/BUILD.gn [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+import("//build_overrides/chip.gni")
+import("${chip_root}/examples/common/pigweed/pigweed_rpcs.gni")
+import("${chip_root}/src/app/chip_data_model.gni")
+
+chip_data_model("switch-common") {
+  zap_file = "switch-app.zap"
+
+  zap_pregenerated_dir = "${chip_root}/zzz_generated/switch-app/zap-generated"
+  is_server = true
```
Comment (dhrishi) [inline examples/switch-app/switch-common/BUILD.gn:23]: What does this `is_server` stand for?

File: examples/switch-app/switch-common/src/ControlCallback.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    if (attributePath.mClusterId != ZCL_ON_OFF_CLUSTER_ID || attributePath.mAttributeId != ZCL_ON_OFF_ATTRIBUTE_ID)
+    {
+        return;
+    }
+    if (size == 0 || value == nullptr)
+    {
+        return;
+    }
+    OnOffCluster clusterClient;
+    clusterClient.Associate(peerDevice, binding.mPeerEndpointId);
```
Comment (dhrishi) [inline examples/switch-app/switch-common/src/ControlCallback.cpp:40]: Any return value given by `Associate`?

File: examples/switch-app/switch-common/switch-app.zap [hunk 8]
```diff
@@ -0,0 +1,4279 @@
...
+      "pathRelativity": "relativeToZap",
+      "path": "../../../src/app/zap-templates/app-templates.json",
+      "version": "chip-v1",
+      "type": "gen-templates-json"
+    }
+  ],
+  "endpointTypes": [
+    {
+      "name": "Anonymous Endpoint Type",
+      "deviceTypeName": "LO-dimmablelight",
```
Comment (dhrishi) [inline examples/switch-app/switch-common/switch-app.zap:35]: TBH, I do not know how difficult it is, but is it possible to keep only the necessary clusters for the switch example?

File: src/app/server/Dnssd.cpp [hunk 9]
```diff
@@ -420,6 +420,11 @@ void DnssdServer::StartServer(chip::Dnssd::CommissioningMode mode)
     {
         ChipLogError(Discovery, "Failed to initialize advertiser: %s", chip::ErrorStr(err));
     }
+    err = Dnssd::Resolver::Instance().Init(&chip::DeviceLayer::InetLayer());
```
Comment (dhrishi) [inline src/app/server/Dnssd.cpp:423]: With this change the resolver will be initialized by default for everyone calling `StartServer`. Is this expected?

File: src/app/binding/BindingTable.cpp [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    }
+    if (device->IsConnected())
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    if (device->IsConnecting())
+    {
+        return CHIP_NO_ERROR;
+    }
+    Dnssd::Resolver::Instance().SetResolverDelegate(&mResolveDelegate);
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:130]: Isn't this clobbering any other resolver delegate that might have been there?

File: src/app/binding/BindingTable.cpp [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    CHIP_ERROR ret = Dnssd::Resolver::Instance().ResolveNodeId(device->GetPeerId(), Inet::IPAddressType::kAny);
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:131]: So if this fails, what happens with `device`?  Who frees it?

File: src/app/binding/BindingTable.cpp [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+        {
+            foundEntry = entry;
+            return false;
+        }
+        return true;
+    });
+    return foundEntry;
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(EndpointId endpoint, ClusterId cluster)
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:154]: Why is this finding an entry, singular?  There can be multiple bindings for the same source endpoint and cluster.  When would this method be used and how would it work in that situation?

File: src/app/binding/BindingTable.cpp [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+    CHIP_ERROR ret = Dnssd::Resolver::Instance().ResolveNodeId(device->GetPeerId(), Inet::IPAddressType::kAny);
+    return ret;
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(NodeId node, FabricIndex fabricIndex)
+{
+    return FindBindingEntry(BuildPeerId(node, fabricIndex));
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(PeerId peer)
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:140]: Why is this a useful operation?  We can have many bindings for a given peer...  This makes it look like there is only one.

If this is meant to be "do we have any binding for the peer", then that's what we should implement.

File: src/app/binding/BindingTable.cpp [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+        }
+        return true;
+    });
+    return foundDevice;
+}
+
+void BindingTable::ResolveDelegate::OnNodeIdResolved(const Dnssd::ResolvedNodeData & nodeData)
+{
+    OperationalDeviceProxy * device = mBindingTable->FindPeerDevice(nodeData.mPeerId);
+    if (nodeData.mNumIPs > 0 && device != nullptr && !device->IsConnected() && !device->IsConnecting())
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:190]: OK, but if `mNumIPs` is 0, then what happens?  Do we just leave the device sitting there?

File: src/app/binding/BindingTable.cpp [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+    {
+        ChipLogProgress(AppServer, "BindingTable: NodeId resolved for 0xdevice %" PRIX64, nodeData.mPeerId.GetNodeId());
+        Transport::PeerAddress addr(nodeData.mAddress[0], Transport::Type::kUdp);
+        addr.SetInterface(nodeData.mInterfaceId);
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:194]: This is wrong.  We should not be using the interface id, except for link-local ipv6 addresses.

See what `DeviceController::ToPeerAddress` does.  We should move that into a utility of some sort....

File: src/app/binding/BindingTable.cpp [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+    });
+    return foundDevice;
+}
+
+void BindingTable::ResolveDelegate::OnNodeIdResolved(const Dnssd::ResolvedNodeData & nodeData)
+{
+    OperationalDeviceProxy * device = mBindingTable->FindPeerDevice(nodeData.mPeerId);
+    if (nodeData.mNumIPs > 0 && device != nullptr && !device->IsConnected() && !device->IsConnecting())
+    {
+        ChipLogProgress(AppServer, "BindingTable: NodeId resolved for 0xdevice %" PRIX64, nodeData.mPeerId.GetNodeId());
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:192]: PRIX64 is not supported on a bunch of our platforms.  Please use ChipLogFormatX64/ChipLogValueX64.

File: src/app/binding/BindingTable.cpp [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+        Transport::PeerAddress addr(nodeData.mAddress[0], Transport::Type::kUdp);
+        addr.SetInterface(nodeData.mInterfaceId);
+        device->UpdateDeviceData(addr, nodeData.mMrpRetryIntervalIdle, nodeData.mMrpRetryIntervalActive);
+        device->Connect(&mBindingTable->mOnDeviceConnected, &mBindingTable->mOnDeviceConnectionFailure);
+    }
+}
+
+void BindingTable::ResolveDelegate::OnNodeIdResolutionFailed(const PeerId & peerId, CHIP_ERROR error)
+{
+    ChipLogError(AppServer, "Failed to resolve device 0x%" PRIX64, peerId.GetNodeId());
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:202]: And what happens with the device?

File: src/app/binding/BindingTable.cpp [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+void BindingTable::OnDeviceConnectedFn(void * context, chip::DeviceProxy * device)
+{
+    ChipLogProgress(AppServer, "Connected to device 0x%" PRIX64, device->GetDeviceId());
+}
+
+void BindingTable::OnDeviceConnectionFailureFn(void * context, NodeId deviceId, CHIP_ERROR error)
+{
+    ChipLogError(AppServer, "Failed to connect to device 0x%" PRIX64, deviceId);
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:212]: Again, what happens with the device?

File: src/app/binding/BindingTable.cpp [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+void MatterBindingAttributeChangeCallback(const chip::app::ConcreteAttributePath & attributePath, uint8_t mask, uint8_t type,
+                                          uint16_t size, uint8_t * value)
+{
+    BindingTableEntry * entry = BindingTable::GetInstance().FindBindingEntry(attributePath.mEndpointId, attributePath.mClusterId);
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:218]: OK, so this is wrong not least because it doesn't look at all entries that match the endpoint/cluster.

But also, this is basically assuming that the time to do something with bindings is when one of our attributes changes?  I guess for on/off that might be enough, ok.

File: src/app/binding/BindingTable.h [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+
+    BindingTableEntry * FindBindingEntry(PeerId peer);
+
+    BindingTableEntry * FindBindingEntry(NodeId node, FabricIndex fabricIndex);
+
+    BindingTableEntry * FindBindingEntry(EndpointId endpoint, ClusterId cluster);
+
+    OperationalDeviceProxy * FindPeerDevice(PeerId peer);
+
+    OperationalDeviceProxy * FindPeerDevice(NodeId node, FabricIndex fabricIndex);
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.h:59]: These all need documenting.  Especially the version of FindBindingEntry that takes endpoint+cluster, and is not documented that this is the _local_ endpoint...

File: src/app/clusters/bindings/bindings.cpp [hunk 21]
```diff
@@ -25,133 +25,63 @@
...
-    if (groupId && nodeId)
-    {
-        return EMBER_BAD_ARGUMENT;
-    }
-
-    binding.clusterId    = clusterId;
-    binding.local        = emberAfCurrentCommand()->apsFrame->destinationEndpoint;
-    binding.networkIndex = 0;
+namespace {
+// TODO: Add peer fabric index to command path
```
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/bindings.cpp:37]: Why are we hardcoding fabric indices at all?

File: src/app/clusters/bindings/bindings.cpp [hunk 22]
```diff
@@ -25,133 +25,63 @@
...
 {
-    auto & nodeId     = commandData.nodeId;
-    auto & groupId    = commandData.groupId;
-    auto & endpointId = commandData.endpointId;
-    auto & clusterId  = commandData.clusterId;
-
-    ChipLogDetail(Zcl, "RX: BindCallback");
-
-    EmberBindingTableEntry bindingEntry;
-    if (prepareBinding(bindingEntry, nodeId, groupId, endpointId, clusterId) != EMBER_SUCCESS)
```
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/bindings.cpp:105]: So here is a question.  why are we rewriting the existing binding table instead of just adding fabric index to it?  It's not a huge deal, but the new implementation is broken for various cases (like groups) that the existing table handles....

File: src/app/binding/BindingTable.cpp [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+ *    limitations under the License.
+ */
+
+#include "app/binding/BindingTable.h"
+
+#include <app-common/zap-generated/attribute-type.h>
+#include <app/server/Server.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+namespace {
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:24]: Why not put this anonymous namespace inside the chip::app namespace below and lose the extra prefixing?

File: src/app/binding/BindingTable.cpp [hunk 24]
```diff
@@ -0,0 +1,237 @@
...
+        device->Connect(&mBindingTable->mOnDeviceConnected, &mBindingTable->mOnDeviceConnectionFailure);
+    }
+}
+
+void BindingTable::ResolveDelegate::OnNodeIdResolutionFailed(const PeerId & peerId, CHIP_ERROR error)
+{
+    ChipLogError(AppServer, "Failed to resolve device 0x%" PRIX64, peerId.GetNodeId());
+}
+
+void BindingTable::OnDeviceConnectedFn(void * context, chip::DeviceProxy * device)
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:205]: Why the various chip:: around here?

File: src/app/server/Dnssd.cpp (hunk 9 above)
Comment (bzbarsky-apple) [inline src/app/server/Dnssd.cpp:423]: Isn't this inside namespace chip?
```suggestion
    err = Dnssd::Resolver::Instance().Init(&DeviceLayer::InetLayer());
```

File: src/app/server/Dnssd.cpp [hunk 25]
```diff
@@ -420,6 +420,11 @@ void DnssdServer::StartServer(chip::Dnssd::CommissioningMode mode)
...
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Discovery, "Failed to initialize resolver: %s", chip::ErrorStr(err));
```
Comment (bzbarsky-apple) [inline src/app/server/Dnssd.cpp:426]:
```suggestion
        ChipLogError(Discovery, "Failed to initialize resolver: %" CHIP_ERROR_FORMAT, err.Format());
```

File: src/app/binding/BindingTable.cpp [hunk 26]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+void MatterBindingAttributeChangeCallback(const chip::app::ConcreteAttributePath & attributePath, uint8_t mask, uint8_t type,
+                                          uint16_t size, uint8_t * value)
+{
+    BindingTableEntry * entry = BindingTable::GetInstance().FindBindingEntry(attributePath.mEndpointId, attributePath.mClusterId);
+    if (entry)
+    {
+        OperationalDeviceProxy * device = BindingTable::GetInstance().FindPeerDevice(entry->mPeerId);
+        if (device && device->IsConnected())
```
Comment (bzbarsky-apple) [inline src/app/binding/BindingTable.cpp:222]: And if it's not connected, maybe the right behavior would be to try connecting it?

File: src/app/util/attribute-table.cpp [hunk 27]
```diff
@@ -614,6 +615,7 @@ EmberAfStatus emAfWriteAttribute(EndpointId endpoint, ClusterId cluster, Attribu
         // Post write attribute callback for all attributes changes, regardless
         // of cluster.
         MatterPostAttributeChangeCallback(attributePath, mask, dataType, emberAfAttributeSize(metadata), data);
+        chip::app::MatterBindingAttributeChangeCallback(attributePath, mask, dataType, emberAfAttributeSize(metadata), data);
```
Comment (bzbarsky-apple) [inline src/app/util/attribute-table.cpp:618]:
```suggestion
        app::MatterBindingAttributeChangeCallback(attributePath, mask, dataType, emberAfAttributeSize(metadata), data);
```
But also, is the expectation that other implementations of MatterBindingAttributeChangeCallback will be provided?  If not, why are we not calling a binding table static method?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/binding/BindingTable.cpp [hunk 28]
```diff
@@ -0,0 +1,237 @@
...
+
+}; // namespace
+
+namespace chip {
+namespace app {
+
+BindingTable BindingTable::sInstance;
+
+BindingTableEntry::BindingTableEntry(EndpointId sourceEndpoint, NodeId peerNodeId, EndpointId peerEndpointId,
+                                     FabricIndex peerFabricIndex, ClusterId clusterId) :
```
[the remark this answers is not in the document]
Comment (gjc13) [reply src/app/binding/BindingTable.cpp:44]: We'll decided group will not be supported in binding for now.

File: src/app/binding/BindingTable.cpp [hunk 29]
```diff
@@ -0,0 +1,237 @@
...
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    if (FindPeerDevice(entry->mPeerId) == nullptr)
+    {
+        DeviceProxyInitParams initParams = { &Server::GetInstance().GetSecureSessionManager(),
+                                             &Server::GetInstance().GetExchangeManager(),
+                                             &Server::GetInstance().GetSessionIDAllocator(),
+                                             GetFabricInfoFromIndex(peerFabricIndex) };
+        OperationalDeviceProxy * device  = mPeerDevices.CreateObject(initParams, entry->mPeerId);
```
[the remark this answers is not in the document]
Comment (gjc13) [reply src/app/binding/BindingTable.cpp:69]: This mechanism is built in chip device controller and is not available on end devices. The mechanism will be modified to reuse one single connection to save resources.

File: src/app/binding/BindingTable.cpp (hunk 10 above)
Comment (gjc13) [reply src/app/binding/BindingTable.cpp:130]: Fortunately we're the only user of the resolver on end devices.

File: src/app/binding/BindingTable.cpp (hunk 11 above)
Comment (gjc13) [reply src/app/binding/BindingTable.cpp:131]: The `device` won't be freed on failures. It's freed by user by unbinding the device.

File: src/app/binding/BindingTable.cpp (hunk 12 above)
Comment (gjc13) [reply src/app/binding/BindingTable.cpp:154]: Removed

File: src/app/binding/BindingTable.cpp (hunk 14 above)
Comment (gjc13) [reply src/app/binding/BindingTable.cpp:190]: Removed

File: src/app/clusters/bindings/bindings.cpp (hunk 21 above)
Comment (gjc13) [reply src/app/clusters/bindings/bindings.cpp:37]: Currently the ZCL code does not pass the fabric in its message context. We know only the endpoint, cluster and the atttribute or command. Adding fabric in context will be in a future PR so as not to make this change too big.

File: src/app/binding/BindingTable.cpp (hunk 28 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/binding/BindingTable.cpp:44]: Who is "we" and why did "we" make that decision while other people are in the middle of trying to stand up group communications?

File: src/app/binding/BindingTable.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/binding/BindingTable.cpp:69]: `CASESessionManager` is not built into the device controller.

File: src/app/binding/BindingTable.cpp (hunk 10 above)
Comment (bzbarsky-apple) [reply src/app/binding/BindingTable.cpp:130]: What guarantees that?  This is a public API; the application code could call it if it wanted to, and nothing tells it not to, as far as I can tell.

File: src/app/binding/BindingTable.cpp (hunk 11 above)
Comment (bzbarsky-apple) [reply src/app/binding/BindingTable.cpp:131]: OK, but it's not connected either.  It just sits there, completely unusable, wasting memory.  Why?

If the intent is to do something in terms of trying to connect it later, etc, that should be documented with TODO comments.

File: src/app/binding/BindingTable.cpp (hunk 14 above)
Comment (bzbarsky-apple) [reply src/app/binding/BindingTable.cpp:190]: Please add TODO comments for any work that is planned but not in this PR yet.  That both aids review and makes it clearer what still needs to be done.

File: src/app/clusters/bindings/bindings.cpp (hunk 21 above)
Comment (bzbarsky-apple) [reply src/app/clusters/bindings/bindings.cpp:37]: The `CommandHandler` knows the fabric and you have one at all the places where kPeerFabricId is being used.  I created https://github.com/project-chip/connectedhomeip/pull/12005 to put a nice API on it, but in the meantime `commandObj->GetExchangeContext()->GetSessionHandle().GetFabricIndex()` will work.

File: src/app/binding/BindingTable.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (gjc13) [reply src/app/binding/BindingTable.cpp:69]: Looks like I've missed the recent changes in `CASESessionManager`. Will re-visit the implementation.

File: src/app/util/attribute-table.cpp (hunk 27 above)
Comment (mrjerryjohns) [inline src/app/util/attribute-table.cpp:618]: Why are making a bespoke call here like this?

File: src/app/binding/BindingTable.h [hunk 30]
```diff
@@ -0,0 +1,104 @@
...
+};
+
+void MatterBindingAttributeChangeCallback(const chip::app::ConcreteAttributePath & attributePath, uint8_t mask, uint8_t type,
+                                          uint16_t size, uint8_t * value);
+
+} // namespace app
+} // namespace chip
+
+// User callback called when a bound cluster is changed. The peer device will be already connected.
+extern void MatterBoundAttributeChangedCallback(const chip::app::BindingTableEntry & binding,
```
Comment (mrjerryjohns) [inline src/app/binding/BindingTable.h:101]: Instead of a free floating function symbol like this, can you please instead define a delegate/callback interface on the Binding cluster implementation that can be used to notify the application of changes?

File: src/app/binding/BindingTable.cpp (hunk 19 above)
Comment (mrjerryjohns) [reply src/app/binding/BindingTable.cpp:218]: > when one of our attributes changes

But...we're on a client, not a server? There are no on/off server attributes on a client right?

File: zzz_generated/switch-app/zap-generated/endpoint_config.h [hunk 31]
```diff
@@ -0,0 +1,865 @@
...
+            { 0x0005, ZAP_TYPE(FABRIC_IDX), 0, ZAP_ATTRIBUTE_MASK(EXTERNAL_STORAGE),                                               \
+              ZAP_EMPTY_DEFAULT() },                                        /* CurrentFabricIndex */                               \
+            { 0xFFFD, ZAP_TYPE(INT16U), 2, 0, ZAP_SIMPLE_DEFAULT(0x0001) }, /* ClusterRevision */                                  \
+                                                                                                                                   \
+            /* Endpoint: 1, Cluster: On/Off (client) */                                                                            \
+            { 0xFFFD, ZAP_TYPE(INT16U), 2, ZAP_ATTRIBUTE_MASK(CLIENT), ZAP_SIMPLE_DEFAULT(2) }, /* ClusterRevision */              \
+                                                                                                                                   \
+            /* Endpoint: 1, Cluster: On/Off (server) */                                                                            \
```
Comment (mrjerryjohns) [inline zzz_generated/switch-app/zap-generated/endpoint_config.h:687]: Why is this here on a client device?

File: src/app/binding/BindingTable.cpp [hunk 32]
```diff
@@ -0,0 +1,237 @@
...
+    {
+        ChipLogProgress(AppServer, "BindingTable: NodeId resolved for 0xdevice %" PRIX64, nodeData.mPeerId.GetNodeId());
+        Transport::PeerAddress addr(nodeData.mAddress[0], Transport::Type::kUdp);
+        addr.SetInterface(nodeData.mInterfaceId);
+        device->UpdateDeviceData(addr, nodeData.mMrpRetryIntervalIdle, nodeData.mMrpRetryIntervalActive);
+        device->Connect(&mBindingTable->mOnDeviceConnected, &mBindingTable->mOnDeviceConnectionFailure);
+    }
+}
+
+void BindingTable::ResolveDelegate::OnNodeIdResolutionFailed(const PeerId & peerId, CHIP_ERROR error)
```
Comment (mrjerryjohns) [inline src/app/binding/BindingTable.cpp:200]: There's an overall lack of any basic error handling here, especially in scenarios where the bound/peer device is not online (or reachable). In those cases, there needs to be some basic level of retrying happening.

File: src/app/binding/BindingTable.h [hunk 33]
```diff
@@ -0,0 +1,104 @@
...
+    GroupId mPeerGroupId; // Destination group, not used
+    ClusterId mClusterId;
+};
+
+class BindingTable
+{
+public:
+    BindingTable();
+
+    CHIP_ERROR Add(EndpointId sourceEndpoint, NodeId peerNodeId, EndpointId peerEndpointId, FabricIndex peerFabricIndex,
```
Comment (mrjerryjohns) [inline src/app/binding/BindingTable.h:45]: What about persistence? What happens when the switch reboots?

**Review by mrjerryjohns — CHANGES_REQUESTED** [review]

File: examples/switch-app/linux/README.md [hunk 34]
```diff
@@ -0,0 +1,63 @@
...
+-   Pair the switch app:
+
+```
+chip-tool pairing ethernet 1235 20202021 3840 127.0.0.1 5540
+```
+
+-   Setup the binding:
+
+```
+chip-tool binding bind 1234 0 1 6 1235 1
```
Comment (markus-becker-tridonic-com) [inline examples/switch-app/linux/README.md:54]: chip-tool (i.e. https://github.com/project-chip/connectedhomeip/blob/master/src/app/zap-templates/zcl/data-model/chip/binding-cluster.xml) is not aligned with spec currently. Current spec has removed the commands `bind`/`unbind`. Bindings are installed by changing the attribute `Binding` (https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Binding-Cluster.adoc#attributes).

**Comment by woody-apple** [comment]
@gjc13 What are the next steps here?

File: examples/switch-app/linux/README.md (hunk 34 above)
Comment (jonsmirl) [reply examples/switch-app/linux/README.md:54]: I'm looking at the Binding cluster specification. I can't see how this is going to work without Bind/Unbind commands. How can you generate sensible error messages if you allow direct manipulation of the attribute? What if you try to write over a slot you don't own? Who allocates the slots?  Who error checks the entities referenced? All of that is possible with bind/unbind commands.

Comment (markus-becker-tridonic-com) [reply examples/switch-app/linux/README.md:54]: Not saying this is sensible, I just noticed the difference. Raise an issue in the spec repo?

**Comment by gjc13** [comment]
Replaced by https://github.com/project-chip/connectedhomeip/pull/12981
